File: apps/macos/Sources/Maumau/UserChannelQuickSetup.swift
import Foundation
import MaumauKit

struct UserChannelQuickSetupLink: Decodable {
    let title: String
    let url: String
}

struct UserChannelQuickSetupGuidance: Decodable {
    let identity: String
    let requirements: [String]
    let setupSteps: [String]
    let artifacts: [String]
    let usage: String?
    let quickLinks: [UserChannelQuickSetupLink]?
}

struct UserChannelQuickSetupField: Decodable {
    let key: String
    let label: String
    let placeholder: String?
    let required: Bool
    let secret: Bool?
    let helpLines: [String]?
}

struct UserChannelQuickSetupCard: Decodable {
    let kind: String
    let sectionTitle: String
    let title: String
    let emptyHeadline: String
    let emptyMessage: String
    let emptyBadge: String
    let buttonTitle: String?
    let existingCredentialNote: String?
    let setupNote: String
    let successMessage: String?
    let waitingBadge: String?
    let waitingMessage: String?
    let linkedBadge: String?
    let qrTitle: String?
    let qrBody: String?
    let pickerSummary: String?
}

struct UserChannelQuickSetupEntry: Decodable {
    let guidance: UserChannelQuickSetupGuidance
    let quickSetup: UserChannelQuickSetupCard
    let fields: [UserChannelQuickSetupField]
}

private struct UserChannelQuickSetupConfig: Decodable {
    let version: Int
    let channelOrder: [String]
    let settingsNote: String
    let channels: [String: UserChannelQuickSetupEntry]
}

enum UserChannelQuickSetupRegistry {
    private static let defaultConfig = Self.loadConfig(resourceName: "user-channel-quick-setup")
    private static let indonesianConfig = Self.loadConfig(resourceName: "user-channel-quick-setup.id")
    private static let fallbackChannelOrder = [
        "whatsapp",
        "telegram",
        "discord",
        "imessage",
        "slack",
        "line",
    ]

    static var channelOrder: [String] {
        let ids = self.defaultConfig?.channelOrder ?? []
        return ids.isEmpty ? self.fallbackChannelOrder : ids
    }

    static var settingsNote: String {
        self.settingsNote(language: .fallback)
    }

    static func settingsNote(language: OnboardingLanguage = .fallback) -> String {
        self.config(for: language)?.settingsNote
            ?? self.defaultConfig?.settingsNote
            ?? "More channels and advanced channel settings live in Settings → Channels."
    }

    static func entry(
        for channelId: String,
        language: OnboardingLanguage = .fallback
    ) -> UserChannelQuickSetupEntry? {
        self.config(for: language)?.channels[channelId]
            ?? self.defaultConfig?.channels[channelId]
    }

    private static func config(for language: OnboardingLanguage) -> UserChannelQuickSetupConfig? {
        switch language {
        case .id:
            self.indonesianConfig ?? self.defaultConfig
        default:
            self.defaultConfig
        }
    }

    private static func loadConfig(resourceName: String) -> UserChannelQuickSetupConfig? {
        guard let url = MaumauKitResources.bundle.url(
            forResource: resourceName,
            withExtension: "json")
        else {
            return nil
        }
        do {
            let data = try Data(contentsOf: url)
            return try JSONDecoder().decode(UserChannelQuickSetupConfig.self, from: data)
        } catch {
            return nil
        }
    }
}

File: src/shared/user-channel-quick-setup.ts
import USER_CHANNEL_QUICK_SETUP_JSON from "../../apps/shared/MaumauKit/Sources/MaumauKit/Resources/user-channel-quick-setup.json" with { type: "json" };
import USER_CHANNEL_QUICK_SETUP_ID_JSON from "../../apps/shared/MaumauKit/Sources/MaumauKit/Resources/user-channel-quick-setup.id.json" with { type: "json" };

export type UserChannelInlineQuickSetupId =
  | "whatsapp"
  | "telegram"
  | "discord"
  | "imessage"
  | "slack"
  | "line";

export type UserChannelQuickSetupGuidanceLink = {
  title: string;
  url: string;
};

export type UserChannelQuickSetupGuidance = {
  identity: string;
  requirements: string[];
  setupSteps: string[];
  artifacts: string[];
  usage?: string;
  quickLinks?: UserChannelQuickSetupGuidanceLink[];
};

export type UserChannelQuickSetupFieldSpec = {
  key: string;
  label: string;
  placeholder?: string;
  required: boolean;
  secret?: boolean;
  helpLines?: string[];
};

export type UserChannelQuickSetupCardSpec = {
  kind: "whatsapp" | "single-secret" | "dual-secret" | "single-text";
  sectionTitle: string;
  title: string;
  emptyHeadline: string;
  emptyMessage: string;
  emptyBadge: string;
  buttonTitle?: string;
  existingCredentialNote?: string;
  setupNote: string;
  successMessage?: string;
  waitingBadge?: string;
  waitingMessage?: string;
  linkedBadge?: string;
  qrTitle?: string;
  qrBody?: string;
  pickerSummary?: string;
};

export type UserChannelQuickSetupEntry = {
  guidance: UserChannelQuickSetupGuidance;
  quickSetup: UserChannelQuickSetupCardSpec;
  fields: UserChannelQuickSetupFieldSpec[];
};

type UserChannelQuickSetupConfig = {
  version: number;
  channelOrder: UserChannelInlineQuickSetupId[];
  settingsNote: string;
  channels: Record<UserChannelInlineQuickSetupId, UserChannelQuickSetupEntry>;
};

export type UserChannelQuickSetupLocale = "en" | "id";

const USER_CHANNEL_QUICK_SETUP_CONFIG =
  USER_CHANNEL_QUICK_SETUP_JSON as UserChannelQuickSetupConfig;
const USER_CHANNEL_QUICK_SETUP_ID_CONFIG =
  USER_CHANNEL_QUICK_SETUP_ID_JSON as UserChannelQuickSetupConfig;

const USER_CHANNEL_QUICK_SETUP_CONFIG_BY_LOCALE = {
  en: USER_CHANNEL_QUICK_SETUP_CONFIG,
  id: USER_CHANNEL_QUICK_SETUP_ID_CONFIG,
} as const satisfies Record<UserChannelQuickSetupLocale, UserChannelQuickSetupConfig>;

export const USER_CHANNEL_INLINE_QUICK_SETUP_IDS = Object.freeze(
  [...USER_CHANNEL_QUICK_SETUP_CONFIG.channelOrder],
) as readonly UserChannelInlineQuickSetupId[];

export function isUserChannelInlineQuickSetupId(
  value: string,
): value is UserChannelInlineQuickSetupId {
  return (USER_CHANNEL_INLINE_QUICK_SETUP_IDS as readonly string[]).includes(value);
}

export function resolveUserChannelQuickSetupLocale(
  locale?: string | null,
): UserChannelQuickSetupLocale {
  const normalized = locale?.trim().toLowerCase() ?? "";
  return normalized.startsWith("id") || normalized.startsWith("in") ? "id" : "en";
}

function quickSetupConfigForLocale(locale?: string | null): UserChannelQuickSetupConfig {
  return USER_CHANNEL_QUICK_SETUP_CONFIG_BY_LOCALE[resolveUserChannelQuickSetupLocale(locale)];
}

export function getUserChannelQuickSetupGuidance(
  channelId: UserChannelInlineQuickSetupId,
): UserChannelQuickSetupGuidance {
  return USER_CHANNEL_QUICK_SETUP_CONFIG.channels[channelId].guidance;
}

export function getUserChannelQuickSetupEntry(
  channelId: UserChannelInlineQuickSetupId,
): UserChannelQuickSetupEntry {
  return USER_CHANNEL_QUICK_SETUP_CONFIG.channels[channelId];
}

export function getLocalizedUserChannelQuickSetupGuidance(
  channelId: UserChannelInlineQuickSetupId,
  locale?: string | null,
): UserChannelQuickSetupGuidance {
  return (
    quickSetupConfigForLocale(locale).channels[channelId]?.guidance ??
    USER_CHANNEL_QUICK_SETUP_CONFIG.channels[channelId].guidance
  );
}

export function getLocalizedUserChannelQuickSetupEntry(
  channelId: UserChannelInlineQuickSetupId,
  locale?: string | null,
): UserChannelQuickSetupEntry {
  return (
    quickSetupConfigForLocale(locale).channels[channelId] ??
    USER_CHANNEL_QUICK_SETUP_CONFIG.channels[channelId]
  );
}

export function getLocalizedUserChannelQuickSetupSettingsNote(locale?: string | null): string {
  return quickSetupConfigForLocale(locale).settingsNote ?? USER_CHANNEL_QUICK_SETUP_CONFIG.settingsNote;
}

export const USER_CHANNEL_QUICK_SETUP_SETTINGS_NOTE =
  USER_CHANNEL_QUICK_SETUP_CONFIG.settingsNote;
